### PR TITLE
handle Route conflicts with HTTPRoute.Matches

### PR DESCRIPTION
**Pull request by lubronzhan**
Fix https://github.com/projectcontour/contour/issues/3608

Now if there is conflicts between HTTPRoute's match, then only the HTTPRoute with
- older creationTimestamp
- alphabetically smaller on `namespace/name`

, will be chosen as the valid HTTPRoute, other HTTPRoute will be marked as conflict condition

TODO:
- [x]  Mark HTTPRoute as conflict condition if needed
- [x]  E2E Test

Signed-off-by: lubronzhan <lubronzhan@gmail.com>

### CODE REVIEW

**Comment by lubronzhan** [comment]
Only failure is context exceeded. I think it's unrelated to the change since the change that affects all route are sorting routes before compute
```
--- FAIL: TestGRPC (0.76s)
    --- FAIL: TestGRPC/StreamRoutes (0.15s)
        server_test.go:200: 
            	Error Trace:	/Users/runner/work/contour/contour/internal/xdscache/v3/server_test.go:298
            	            				/Users/runner/work/contour/contour/internal/xdscache/v3/server_test.go:200
            	            				/Users/runner/work/contour/contour/internal/xdscache/v3/server_test.go:276
            	Error:      	Received unexpected error:
            	            	rpc error: code = DeadlineExceeded desc = context deadline exceeded
            	Test:       	TestGRPC/StreamRoutes
```

File: internal/dag/status_test.go
Comment (lubronzhan) [inline internal/dag/status_test.go:5732]: My question here, the conflicted route is still shown attached. Is that ok?

**Comment by lubronzhan** [comment]
hi @sunjayBhatia could you take a look

File: internal/dag/status_test.go
Comment (sunjayBhatia) [reply internal/dag/status_test.go:5733]: Yep, that is correct:

```
	// Successful attachment of a Route to a Listener is based solely on the
	// combination of the AllowedRoutes field on the corresponding Listener
	// and the Route's ParentRefs field. A Route is successfully attached to
	// a Listener when it is selected by the Listener's AllowedRoutes field
	// AND the Route has a valid ParentRef selecting the whole Gateway
	// resource or a specific Listener as a parent resource (more detail on
	// attachment semantics can be found in the documentation on the various
	// Route kinds ParentRefs fields). Listener or Route status does not impact
	// successful attachment, i.e. the AttachedRoutes field count MUST be set
	// for Listeners with condition Accepted: false and MUST count successfully
	// attached Routes that may themselves have Accepted: false conditions.
```

From: https://github.com/kubernetes-sigs/gateway-api/blob/72f679c7c67228f5b4340e0c2f50cc04b36945b1/apis/v1/gateway_types.go#L849-L859

File: internal/dag/gatewayapi_processor_test.go
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor_test.go:775]:
```suggestion
	time1 := meta_v1.Date(2021, time.Month(2), 21, 1, 10, 30, 0, time.UTC)
```

nit

File: internal/dag/gatewayapi_processor.go
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:1438]: I think we might want to reconsider what happens when a route has at least one rule that conflicts with another but has some rules that are still valid (so as not to drop *all* traffic for this route)

we may want to not drop all the route rules and use the `PartiallyInvalid: True` Condition: https://github.com/kubernetes-sigs/gateway-api/blob/72f679c7c67228f5b4340e0c2f50cc04b36945b1/apis/v1/shared_types.go#L435

cc @projectcontour/maintainers @projectcontour/contour-reviewers

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]: Yeah i remember i read this similar discussion somewhere, do we expect rest of the route to be valid if just some of them returns 500

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]: But the route object we are  trying to drop here is already single route rule I assume? the matching criteria is constructed by `	PathMatchCondition&&HeaderMatchConditions&&QueryParamMatchConditions` under this route

Or do I miss anything?

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]: ~~But the route object we are  trying to drop here is already single route rule I assume? the matching criteria is constructed by `	PathMatchCondition&&HeaderMatchConditions&&QueryParamMatchConditions` under this route~~

Oh yeah the logic I have now is to mark Route as invalid and not adding all routes under it to hosts.

Do we want to change it to partiallyInvalid? cc @skriss . I think it make snse

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]: Ok create an issue here https://github.com/projectcontour/contour/issues/6227
Also let me update the logic then

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]:
```
	// 1) Drop Rule(s): With this approach, implementations will drop the
	//    invalid Route Rule(s) until they are fully valid again. The message
	//    for this condition MUST start with the prefix "Dropped Rule" and
	//    include information about which Rules have been dropped. In this
	//    state, the "Accepted" condition MUST be set to "True" with the latest
	//    generation of the resource.
```

https://github.com/kubernetes-sigs/gateway-api/blob/72f679c7c67228f5b4340e0c2f50cc04b36945b1/apis/v1/shared_types.go#L408C1-L413C35

The question i have here is it says about dropping `Rule`, so do we consider conflict at Rule level? If there is one route under it has conflict, we drop the rule?

Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor.go:1438]: It sounds like we might be switching the definition of "route" and "rule" here?

In the upstream spec "route" should refer to the whole k8s resource (HTTPRoute) vs. "rule" is https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteRule

So if there is conflict with a single "rule" under a "route", only that rule would be dropped and other parts of the route should still work

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]: I was talking about this [Route object](https://github.com/lubronzhan/contour/blob/topic/lubron/fix-3608/internal/dag/dag.go#L277-L377). 
Should I call it cluster route? It's fetched from [here](https://github.com/projectcontour/contour/blob/main/internal/dag/gatewayapi_processor.go#L1396-L1427) 

What's the difference between this Route to the HttpRoute object?

Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor.go:1438]: ah gotcha, yeah the `dag.Route` type more matches up the the `HTTPRouteRule` type (the `dag.VirtualHosts` and the `dag.Routes` under them basically come from a `HTTPRoute` and the Gateway `Listeners`, including the hostnames configured on the Route and Gateway)

so if we're processing a `HTTPRouteRule`+hostname (from the `HTTPRoute` and/or `Listener`) that has match conditions that conflict with an existing `dag.VirtualHost`+`dag.Route` that has already been added to the DAG (and therefore a `HTTPRouteRule` from another `HTTPRoute`) then we should not add it, set status, etc.

one complication I forgot to mention is that we've sort of implicitly allowed the processors that run later in the DAG build process to overwrite `dag.Routes` etc. from older ones (Gateway API can overwrite HTTPProxy can overwrite Ingress), if we make this change without any awareness of what can overwrite what, then we're kinda reversing this for the Gateway API processor (it won't be able to have precedence over other kinds)

Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor.go:1438]: was thinking something like this might work:

```diff
diff --git a/internal/dag/dag.go b/internal/dag/dag.go
index b2bf912c..7710483d 100644
--- a/internal/dag/dag.go
+++ b/internal/dag/dag.go
@@ -769,6 +769,21 @@ func (v *VirtualHost) HasConflictRoute(route *Route) bool {
 	return false
 }
 
+func (v *VirtualHost) AddRouteIfNoConflict(route *Route) bool {
+	if v.Routes == nil {
+		v.Routes = make(map[string]*Route)
+	}
+	existingRoute, found := v.Routes[conditionsToString(route)]
+	if found {
+		return false
+	}
+	if existingRoute.Kind != route.Kind {
+		v.Routes[conditionsToString(route)] = route
+		return true
+	}
+	return false
+}
+
 func conditionsToString(r *Route) string {
 	s := []string{r.PathMatchCondition.String()}
 	for _, cond := range r.HeaderMatchConditions {
diff --git a/internal/dag/gatewayapi_processor.go b/internal/dag/gatewayapi_processor.go
index de518151..e2118fb5 100644
--- a/internal/dag/gatewayapi_processor.go
+++ b/internal/dag/gatewayapi_processor.go
@@ -1434,27 +1434,27 @@ func (p *GatewayAPIProcessor) computeHTTPRouteForListener(
 	}
 
 	// check all the routes at once in case there is conflict
-	if p.hasConflictRoute(listener, hosts, routes) {
-		// skip adding the route to svhost or vhost since it's marked as conflict route
-		routeAccessor.AddCondition(
-			gatewayapi_v1.RouteConditionAccepted,
-			meta_v1.ConditionFalse,
-			status.ReasonRouteRuleMatchConflict,
-			"HTTPRoute's Match has conflict with other HTTPRoute's Match",
-		)
-	} else {
-		// Add each route to the relevant vhost(s)/svhosts(s).
-		for host := range hosts {
-			for _, route := range routes {
-				switch {
-				case listener.tlsSecret != nil:
-					svhost := p.dag.EnsureSecureVirtualHost(listener.dagListenerName, host)
-					svhost.Secret = listener.tlsSecret
-					svhost.AddRoute(route)
-				default:
-					vhost := p.dag.EnsureVirtualHost(listener.dagListenerName, host)
-					vhost.AddRoute(route)
-				}
+	// Add each route to the relevant vhost(s)/svhosts(s).
+	for host := range hosts {
+		for _, route := range routes {
+			routeAdded := false
+			switch {
+			case listener.tlsSecret != nil:
+				svhost := p.dag.EnsureSecureVirtualHost(listener.dagListenerName, host)
+				svhost.Secret = listener.tlsSecret
+				routeAdded = svhost.AddRouteIfNoConflict(route)
+			default:
+				vhost := p.dag.EnsureVirtualHost(listener.dagListenerName, host)
+				routeAdded = vhost.AddRouteIfNoConflict(route)
+			}
+			if !routeAdded {
+				// skip adding the route to svhost or vhost since it's marked as conflict route
+				routeAccessor.AddCondition(
+					gatewayapi_v1.RouteConditionAccepted,
+					meta_v1.ConditionFalse,
+					status.ReasonRouteRuleMatchConflict,
+					"HTTPRoute's Match has conflict with other HTTPRoute's Match",
+				)
 			}
 		}
 	}
```

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]: Thanks for the detail, I checked the code again, one `dag.Route` is generated from one of the `Match` under `HTTPRouteRule`, so one rule could have several matches, thus several `dag.Route`, if we find single `Match` or `dag.Route`, under a `HTTPRouteRule` has conflict with other `dag.Route` with other `dag.Route`, then we should drop the whole `HTTPRouteRule`, not just the single `dag.Route` I assume.

Now I understand more about the relationship between `Route`, `dag.Route` and `Rule`, thanks

Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1438]: MM not sure how do we indicate which rule has been dropped, since it doesn't have name under it, and it contains many matches. I guess I'll skip adding that info to the condition for now.
```
"Dropped Rule" and
	//    include information about which Rules have been dropped. 

```

File: internal/dag/dag_test.go
Comment (sunjayBhatia) [reply internal/dag/dag_test.go:339]: I'm realizing we've created another edge case with the relationship between HTTPRoute and GRPCRoute, we implement GRPCRoute service/method matching with HTTP path matches so you could conceivably have a conflict between a HTTPRoute path match and GRPCRoute service/method match

If we leave the implementation as-is, GRPCRoutes could override HTTPRoutes by virtue of being processed later

(also we should probably do the same conflict logic detection for GRPCRoutes, but seems ok to do in a follow up if that makes this too big of a PR)

**Review by sunjayBhatia — COMMENTED** [review]
cc @projectcontour/maintainers for others thoughts on this change

File: internal/dag/dag_test.go
Comment (lubronzhan) [reply internal/dag/dag_test.go:339]: Thanks. Good catch. Then I can quickly create a PR for GRPCRoute once this is merged. Since the content would be very similar

**Comment by lubronzhan** [comment]
Hi @projectcontour/maintainers any chance to get this merged?

File: test/e2e/gateway/http_route_partially_conflict_match_test.go
Comment (skriss) [inline test/e2e/gateway/http_route_partially_conflict_match_test.go:103]: I think you can combine these:
```suggestion
		f.CreateHTTPRouteAndWaitFor(route2, func(...) {
		        return e2e.HTTPRoutePartiallyAccepted(route2) && e2e.HTTPRouteAccepted(route2)
		})
```

File: test/e2e/gatewayapi_predicates.go
Comment (skriss) [inline test/e2e/gatewayapi_predicates.go:144]: I would probably rename this to `HTTPRoutePartiallyInvalid` since that's the condition it's looking for, it's a little more clear that way what it's doing.

File: test/e2e/gateway/http_route_conflict_match_test.go
Comment (skriss) [inline test/e2e/gateway/http_route_conflict_match_test.go:59]: We should be checking the return values here to ensure it did eventually get accepted:
```suggestion
		_, ok := f.CreateHTTPRouteAndWaitFor(route1, e2e.HTTPRouteAccepted)
		require.True(f.T(), ok)
```

(Now seeing tons of places where we're not doing this in existing tests, I'll look at addressing that separately)

File: test/e2e/gateway/http_route_conflict_match_test.go
Comment (skriss) [inline test/e2e/gateway/http_route_conflict_match_test.go:84]: require return val to be true

File: test/e2e/gateway/http_route_partially_conflict_match_test.go
Comment (skriss) [inline test/e2e/gateway/http_route_partially_conflict_match_test.go:64]: require return val to be true

File: test/e2e/gateway/http_route_partially_conflict_match_test.go
Comment (skriss) [inline test/e2e/gateway/http_route_partially_conflict_match_test.go:96]: require return val to be true

File: test/e2e/gatewayapi_predicates.go
Comment (skriss) [inline test/e2e/gatewayapi_predicates.go:151]:
```suggestion
			return true
```

File: test/e2e/gatewayapi_predicates.go
Comment (skriss) [inline test/e2e/gatewayapi_predicates.go:134]:
```suggestion
			return true
```

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks @lubronzhan, I think this change makes sense though it would definitely be nice to have an upstream Gateway API conformance test to verify against, maybe we can contribute that separately

**Comment by lubronzhan** [comment]
> Thanks @lubronzhan, I think this change makes sense though it would definitely be nice to have an upstream Gateway API conformance test to verify against, maybe we can contribute that separately

Do you have any reference. Just saw we are referencing the conformance test [here](https://github.com/projectcontour/contour/blob/main/test/conformance/gatewayapi/gateway_conformance_test.go#L143)

**Comment by skriss** [comment]
> > Thanks @lubronzhan, I think this change makes sense though it would definitely be nice to have an upstream Gateway API conformance test to verify against, maybe we can contribute that separately
> 
> Do you have any reference. Just saw we are referencing the conformance test [here](https://github.com/projectcontour/contour/blob/main/test/conformance/gatewayapi/gateway_conformance_test.go#L143)
> 
> Happy to put it into different PR

Yeah it would be in the Gateway API repo, the conformance tests are here: https://github.com/kubernetes-sigs/gateway-api/tree/main/conformance/tests

You can see some of the past PRs to add tests here: https://github.com/kubernetes-sigs/gateway-api/pulls?q=is%3Apr+label%3Aarea%2Fconformance

Happy to pair/provide further pointers if needed

File: internal/dag/gatewayapi_processor.go
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1447]: One case I'm not totally clear on is: what if a rule in a given HTTPRoute has multiple matches defined, and only _one_ of those matches conflicts with another HTTPRoute? Should the entire rule be dropped, or only the match that conflicted? (matches within a rule are logically OR'ed)

https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.RouteConditionType docs for PartiallyInvalid seem to imply that the entire rule should be dropped, but it's not crystal clear from the spec. I think this PR's implementation is good for now (dropping the entire rule), but this is an example of something that could be clarified and encoded in the upstream conformance test.

**Review by skriss — APPROVED** [review]

**Comment by lubronzhan** [comment]
> > > Thanks @lubronzhan, I think this change makes sense though it would definitely be nice to have an upstream Gateway API conformance test to verify against, maybe we can contribute that separately
> > 
> > 
> > Do you have any reference. Just saw we are referencing the conformance test [here](https://github.com/projectcontour/contour/blob/main/test/conformance/gatewayapi/gateway_conformance_test.go?rgh-link-date=2024-04-16T16%3A33%3A50Z#L143)
> > Happy to put it into different PR
> 
> Yeah it would be in the Gateway API repo, the conformance tests are here: [kubernetes-sigs/gateway-api@`main`/conformance/tests](https://github.com/kubernetes-sigs/gateway-api/tree/main/conformance/tests?rgh-link-date=2024-04-16T16%3A33%3A50Z)
> 
> You can see some of the past PRs to add tests here: [kubernetes-sigs/gateway-api/pulls (label:area/conformance)](https://github.com/kubernetes-sigs/gateway-api/pulls?q=is%3Apr+label%3Aarea%2Fconformance+sort%3Aupdated-desc)
> 
> Happy to pair/provide further pointers if needed

Just found out that I did created an issue for tracking conformance test https://github.com/projectcontour/contour/issues/6227

Thanks for the pointers, will create one for gateway-api

File: internal/dag/gatewayapi_processor.go
Comment (lubronzhan) [reply internal/dag/gatewayapi_processor.go:1447]: Sunjay also mentioned [here](https://github.com/projectcontour/contour/pull/6188#discussion_r1505010896), in upstream, [it does indicate](https://github.com/kubernetes-sigs/gateway-api/blob/72f679c7c67228f5b4340e0c2f50cc04b36945b1/apis/v1/shared_types.go#L408C1-L413C35) that we should drop at the rule level, if one rule is conflict, just drop it and leave other rules
>// 1) Drop Rule(s): With this approach, implementations will drop the
	//    invalid Route Rule(s) until they are fully valid again. The message
	//    for this condition MUST start with the prefix "Dropped Rule" and
	//    include information about which Rules have been dropped. In this
	//    state, the "Accepted" condition MUST be set to "True" with the latest
	//    generation of the resource.

**Comment by skriss** [comment]
:+1: I'm going to go ahead and merge this now so it doesn't run into any conflicts, can always circle back and address any post-merge feedback if needed before release. Thanks again @lubronzhan!

File: internal/dag/dag_test.go
Comment (lubronzhan) [reply internal/dag/dag_test.go]: >I'm realizing we've created another edge case with the relationship between HTTPRoute and GRPCRoute, we implement GRPCRoute service/method matching with HTTP path matches so you could conceivably have a conflict between a HTTPRoute path match and GRPCRoute service/method match

Hi @sunjayBhatia sorry I think I missed this comment, I did compare the route kind when checking the route conflict. Do you think that's enough? https://github.com/projectcontour/contour/blob/main/internal/dag/dag.go#L767

**Pull request merged**
